### PR TITLE
Additional groups lookup

### PR DESCRIPTION
**Pull request by mrunalp**
This PR modifies AdditionalGroups to be a string array and the lookup to translate to group ids is performed in the container as discussed in https://github.com/docker/docker/pull/10717

Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

### CODE REVIEW

**Comment by mrunalp** [comment]
@tianon PTAL :)

File: user/user.go [hunk 1]
```diff
@@ -348,3 +348,52 @@ func GetExecUser(userSpec string, defaults *ExecUser, passwd, group io.Reader) (
...
+// given file path and uses that data as the arguments to GetAdditionalGroups.
+func GetAdditionalGroupsPath(additionalGroups []string, groupPath string) ([]int, error) {
+	var groupIds []int
+
+	for _, ag := range additionalGroups {
+		groupReader, err := os.Open(groupPath)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to open group file: %v", err)
+		}
+		defer groupReader.Close()
```
Comment (dqminh) [inline user/user.go]: i think `groupPath` can be opened once and reused after.

File: user/user.go [hunk 2]
```diff
@@ -348,3 +348,52 @@ func GetExecUser(userSpec string, defaults *ExecUser, passwd, group io.Reader) (
...
+
+		groupId, err := GetAdditionalGroup(ag, groupReader)
```
Comment (dqminh) [inline user/user.go]: Should `GetAdditionalGroup` accepts a list of groups too so it doesnt have to re-parse the group file for each group ? Or maybe we can merge `GetAdditionalGroup` into this function so it's more obvious that we will only read the group file once.

File: user/user.go [hunk 3]
```diff
@@ -348,3 +348,52 @@ func GetExecUser(userSpec string, defaults *ExecUser, passwd, group io.Reader) (
...
+		if err != nil {
+			return nil, err
+		}
+		groupIds = append(groupIds, groupId)
+	}
+
+	return groupIds, nil
+}
```
Comment (dqminh) [inline user/user.go]: I think we should also have an unit test for this.

Comment (mrunalp) [inline user/user.go]: Yes, will add one.

File: user/user.go (hunk 2 above)
Comment (mrunalp) [inline user/user.go]: I went down that path but realized that the existing filter won't work for it. I can however create a new filter that checks for multiple groups at the same time so we can avoid multiple passes through the group reader.

Comment (dqminh) [inline user/user.go]: I think adding a filter that works with multiple groups sounds good.

**Comment by LK4D4** [comment]
ping @mrunalp :)
Would be nice to have docker/docker#10717 in next release

**Comment by mrunalp** [comment]
Am out sick :( hopefully next week

**Comment by dqminh** [comment]
@mrunalp I have some free time ! I can send you some improvements of this if you dont mind.

**Comment by mrunalp** [comment]
Sure, go for it. Thanks.

**Comment by LK4D4** [comment]
Closing in favor of #603

**Pull request closed without merging**